src/mongocrypt-marking.c

```diff
@@ -512,120 +550,134 @@ _mongocrypt_fle2_placeholder_common (_mongocrypt_key_broker_t *kb,
+      goto fail;
    }
 
    _mongocrypt_buffer_copy_to (&placeholder->index_key_id,
-                               &payload.indexKeyId);          // u
-   payload.valueType = bson_iter_type (&placeholder->v_iter); // t
+                               &out->indexKeyId); // u
+   out->valueType = bson_iter_type (value_iter);  // t
 
-   // v := UserKeyId + EncryptAEAD(UserKey, value)
+   // v := UserKeyId + EncryptCTRAEAD(UserKey, value)
    {
-      _mongocrypt_buffer_t tmp[2] = {placeholder->user_key_id, {0}};
-      if (!_fle2_placeholder_aead_encrypt (
-             kb, &placeholder->user_key_id, &value, &tmp[1], status)) {
+      _mongocrypt_buffer_t ciphertext = {0};
+      if (!_fle2_placeholder_aes_ctr_aead_encrypt (
+             kb, &ciphertext, &placeholder->user_key_id, &value, status)) {
          goto fail;
       }
-      bool ok = _mongocrypt_buffer_concat (&payload.value, tmp, 2);
-      _mongocrypt_buffer_cleanup (&tmp[1]);
+      const _mongocrypt_buffer_t v[2] = {placeholder->user_key_id, ciphertext};
+      const bool ok = _mongocrypt_buffer_concat (&out->value, v, 2);
+      _mongocrypt_buffer_cleanup (&ciphertext);
       if (!ok) {
          goto fail;
       }
    }
 
    // e := ServerDataEncryptionLevel1Token
-   {
-      mc_ServerDataEncryptionLevel1Token_t *serverToken =
-         mc_ServerDataEncryptionLevel1Token_new (
-            crypto, &common.tokenKey, status);
-      if (!serverToken) {
-         goto fail;
-      }
-      _mongocrypt_buffer_copy_to (
-         mc_ServerDataEncryptionLevel1Token_get (serverToken),
-         &payload.serverEncryptionToken);
-      mc_ServerDataEncryptionLevel1Token_destroy (serverToken);
+   _mongocrypt_buffer_copy_to (mc_ServerDataEncryptionLevel1Token_get (
+                                  common->serverDataEncryptionLevel1Token),
+                               &out->serverEncryptionToken);
+
+   res = true;
+fail:
+   _mongocrypt_buffer_cleanup (&value);
+   return res;
+}
+
+/**
+ * Payload subtype 4: FLE2InsertUpdatePayload
+ *
+ * {d: EDC, s: ESC, c: ECC, p: encToken, v: value, e: serverToken}
+ */
+static bool
+_mongocrypt_fle2_placeholder_to_insert_update_ciphertext (
+   _mongocrypt_key_broker_t *kb,
+   _mongocrypt_marking_t *marking,
+   _mongocrypt_ciphertext_t *ciphertext,
+   mongocrypt_status_t *status)
+{
+   mc_FLE2EncryptionPlaceholder_t *placeholder = &marking->fle2;
+   _FLE2EncryptedPayloadCommon_t common = {{0}};
+   mc_FLE2InsertUpdatePayload_t payload;
+   mc_FLE2InsertUpdatePayload_init (&payload);
+   bool res = false;
+
+   BSON_ASSERT (marking->type == MONGOCRYPT_MARKING_FLE2_ENCRYPTION);
+
+   int64_t contentionFactor = 0;
+   if (!_mongocrypt_fle2_placeholder_to_insert_update_common (
+          kb,
+          &payload,
+          &contentionFactor,
+          &common,
+          placeholder,
+          &placeholder->v_iter,
+          status)) {
+      goto fail;
    }
 
+   _mongocrypt_ciphertext_init (ciphertext);
```

Comment:
```suggestion
```

Current callers of `_mongocrypt_marking_to_ciphertext` call `_mongocrypt_ciphertext_init` on the passed `ciphertext`. Suggest removing the `_mongocrypt_ciphertext_init` to not attempt to support uninitialized `ciphertext` from being passed.

Suggest adding a comment to `mongocrypt-marking-private.h` to communicate the expectation:
```c
// Callers are expected to initialize `ciphertext` with _mongocrypt_ciphertext_init before calling.
bool
_mongocrypt_marking_to_ciphertext (void *ctx,
                                   _mongocrypt_marking_t *marking,
                                   _mongocrypt_ciphertext_t *ciphertext,
                                   mongocrypt_status_t *status)
   MONGOCRYPT_WARN_UNUSED_RESULT;
```

Reply:
Got it. Yeah, I see the caller now.  Added comment and removed inits here and throughout callees of _mongocrypt_marking_to_ciphertext.